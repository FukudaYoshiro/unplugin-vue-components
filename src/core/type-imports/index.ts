import type { TypeImport } from '../../types'

export const TypeImportPresets: TypeImport[] = [
  {
    from: 'vue-router',
    names: [
      'RouterView',
      'RouterLink',
    ],
  },
  {
    from: 'vue-starport',
    names: [
      'Starport',
      'StarportCarrier',
    ],
  },
]
