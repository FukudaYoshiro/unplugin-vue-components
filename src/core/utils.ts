import { parse } from 'path'
import minimatch from 'minimatch'
import resolve from 'resolve'
import { slash, toArray } from '@antfu/utils'
import {
  getPackageInfo,
  isPackageExists,
} from 'local-pkg'
import type { ComponentInfo, ImportInfo, ResolvedOptions } from '../types'
import type { Context } from './context'
import { DISABLE_COMMENT } from './constants'

export const isSSR = Boolean(process.env.SSR || process.env.SSG || process.env.VITE_SSR || process.env.VITE_SSG)

export interface ResolveComponent {
  filename: string
  namespace?: string
}

export function pascalCase(str: string) {
  return capitalize(camelCase(str))
}

export function camelCase(str: string) {
  return str.replace(/-(\w)/g, (_, c) => (c ? c.toUpperCase() : ''))
}

export function kebabCase(key: string) {
  const result = key.replace(/([A-Z])/g, ' $1').trim()
  return result.split(' ').join('-').toLowerCase()
}

export function capitalize(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1)
}

export function parseId(id: string) {
  const index = id.indexOf('?')
  if (index < 0) {
    return { path: id, query: {} }
  }
  else {
    const query = Object.fromEntries(new URLSearchParams(id.slice(index)) as any)
    return {
      path: id.slice(0, index),
      query,
    }
  }
}

export function isEmpty(value: any) {
  if (!value || value === null || value === undefined || (Array.isArray(value) && Object.keys(value).length <= 0))
    return true

  else
    return false
}

export function matchGlobs(filepath: string, globs: string[]) {
  for (const glob of globs) {
    if (minimatch(slash(filepath), glob))
      return true
  }
  return false
}

export function getTransformedPath(path: string, ctx: Context): string {
  if (ctx.options.importPathTransform) {
    const result = ctx.options.importPathTransform(path)
    if (result != null)
      path = result
  }

  return path
}

export function stringifyImport(info: ImportInfo | string) {
  if (typeof info === 'string')
    return `import '${info}'`
  if (!info.name)
    return `import '${info.path}'`
  else if (info.importName)
    return `import { ${info.importName} as ${info.name} } from '${info.path}'`
  else
    return `import ${info.name} from '${info.path}'`
}

export function stringifyComponentImport({ name, path, importName, sideEffects }: ComponentInfo, ctx: Context) {
  path = getTransformedPath(path, ctx)

  const imports = [
    stringifyImport({ name, path, importName }),
  ]

  if (sideEffects)
    toArray(sideEffects).forEach(i => imports.push(stringifyImport(i)))

  return imports.join(';')
}

export function getNameFromFilePath(filePath: string, options: ResolvedOptions): string {
  const { resolvedDirs, directoryAsNamespace, globalNamespaces } = options

  const parsedFilePath = parse(slash(filePath))

  let strippedPath = ''

  // remove include directories from filepath
  for (const dir of resolvedDirs) {
    if (parsedFilePath.dir.startsWith(dir)) {
      strippedPath = parsedFilePath.dir.slice(dir.length)
      break
    }
  }

  let folders = strippedPath.slice(1).split('/').filter(Boolean)
  let filename = parsedFilePath.name

  // set parent directory as filename if it is index
  if (filename === 'index' && !directoryAsNamespace) {
    filename = `${folders.slice(-1)[0]}`
    return filename
  }

  if (directoryAsNamespace) {
    // remove namesspaces from folder names
    if (globalNamespaces.some((name: string) => folders.includes(name)))
      folders = folders.filter(f => !globalNamespaces.includes(f))

    if (filename.toLowerCase() === 'index')
      filename = ''

    if (!isEmpty(folders)) {
      // add folders to filename
      filename = [...folders, filename].filter(Boolean).join('-')
    }

    return filename
  }

  return filename
}

export function resolveAlias(filepath: string, alias: any) {
  const result = filepath
  if (Array.isArray(alias)) {
    for (const { find, replacement } of alias)
      result.replace(find, replacement)
  }
  return result
}

export async function getPkgVersion(pkgName: string, defaultVersion: string): Promise<string> {
  try {
    const isExist = isPackageExists(pkgName)
    if (isExist) {
      const pkg = await getPackageInfo(pkgName)
      return pkg?.version ?? defaultVersion
    }
    else {
      return defaultVersion
    }
  }
  catch (err) {
    console.error(err)
    return defaultVersion
  }
}

export function shouldTransform(code: string) {
  if (code.includes(DISABLE_COMMENT))
    return false
  return true
}

export function resolveImportPath(importName: string): string | undefined {
  return resolve.sync(importName, {
    preserveSymlinks: false,
  })
}
