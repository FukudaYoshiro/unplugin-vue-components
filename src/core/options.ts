import { join, resolve } from 'path'
import { slash, toArray } from '@antfu/utils'
import { isPackageExists } from 'local-pkg'
import type { ComponentResolver, ComponentResolverObject, Options, ResolvedOptions } from '../types'
import { detectTypeImports } from './type-imports/detect'

export const defaultOptions: Omit<Required<Options>, 'include' | 'exclude' | 'transformer' | 'globs' | 'directives' | 'types'> = {
  dirs: 'src/components',
  extensions: 'vue',
  deep: true,
  dts: isPackageExists('typescript'),

  directoryAsNamespace: false,
  globalNamespaces: [],

  resolvers: [],

  importPathTransform: v => v,

  allowOverrides: false,
}

function normalizeResolvers(resolvers: (ComponentResolver | ComponentResolver[])[]): ComponentResolverObject[] {
  return toArray(resolvers).flat().map(r => typeof r === 'function' ? { resolve: r, type: 'component' } : r)
}

export function resolveOptions(options: Options, root: string): ResolvedOptions {
  const resolved = Object.assign({}, defaultOptions, options) as ResolvedOptions
  resolved.resolvers = normalizeResolvers(resolved.resolvers)
  resolved.extensions = toArray(resolved.extensions)

  if (resolved.globs) {
    resolved.globs = toArray(resolved.globs).map((glob: string) => slash(resolve(root, glob)))
    resolved.resolvedDirs = []
  }
  else {
    const extsGlob = resolved.extensions.length === 1
      ? resolved.extensions
      : `{${resolved.extensions.join(',')}}`

    resolved.dirs = toArray(resolved.dirs)
    resolved.resolvedDirs = resolved.dirs.map(i => slash(resolve(root, i)))

    resolved.globs = resolved.resolvedDirs.map(i => resolved.deep
      ? slash(join(i, `**/*.${extsGlob}`))
      : slash(join(i, `*.${extsGlob}`)),
    )

    if (!resolved.extensions.length)
      throw new Error('[unplugin-vue-components] `extensions` option is required to search for components')
  }

  resolved.dts = !resolved.dts
    ? false
    : resolve(
      root,
      typeof resolved.dts === 'string'
        ? resolved.dts
        : 'components.d.ts',
    )

  if (!resolved.types && resolved.dts)
    resolved.types = detectTypeImports()
  resolved.types = resolved.types || []

  resolved.root = root
  resolved.transformer = options.transformer || getVueVersion() || 'vue3'
  resolved.directives = (typeof options.directives === 'boolean')
    ? options.directives
    : !resolved.resolvers.some(i => i.type === 'directive')
      ? false
      : getVueVersion() === 'vue3'

  return resolved
}

function getVueVersion() {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const vue = require('vue')
    const version = vue?.default?.version || vue?.version || '3'
    return version.startsWith('2.') ? 'vue2' : 'vue3'
  }
  catch {
    return null
  }
}
