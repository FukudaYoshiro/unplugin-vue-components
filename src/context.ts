import { relative } from 'path'
import Debug from 'debug'
import { ResolvedConfig, UpdatePayload, ViteDevServer } from 'vite'
import { throttle, toArray, slash } from '@antfu/utils'
import { Options, ComponentInfo, ResolvedOptions } from './types'
import { pascalCase, getNameFromFilePath, resolveAlias, resolveOptions, matchGlobs } from './utils'
import { searchComponents } from './fs/glob'
import { generateDeclaration } from './declaration'

const debug = {
  components: Debug('vite-plugin-components:context:components'),
  search: Debug('vite-plugin-components:context:search'),
  hmr: Debug('vite-plugin-components:context:hmr'),
  decleration: Debug('vite-plugin-components:decleration'),
}

export class Context {
  readonly options: ResolvedOptions

  private _componentPaths = new Set<string>()
  private _componentNameMap: Record<string, ComponentInfo> = {}
  private _componentUsageMap: Record<string, Set<string>> = {}
  private _componentCustomMap: Record<string, ComponentInfo> = {}
  private _server: ViteDevServer | undefined

  constructor(
    options: Options,
    public readonly viteConfig: ResolvedConfig,
  ) {
    this.options = resolveOptions(options, viteConfig)
    this.generateDeclaration = throttle(500, false, this.generateDeclaration.bind(this))
  }

  get root() {
    return this.viteConfig.root
  }

  setServer(server: ViteDevServer) {
    this._server = server

    const { globs, dirs } = this.options

    server.watcher.add(dirs)
    server.watcher
      .on('unlink', (path) => {
        if (!matchGlobs(path, globs))
          return
        this.removeComponents(path)
        this.onUpdate(path)
      })
    server.watcher
      .on('add', (path) => {
        if (!matchGlobs(path, globs))
          return
        this.addComponents(path)
        this.onUpdate(path)
      })
  }

  /**
   * Record the usage of components
   * @param path
   * @param paths paths of used components
   */
  updateUsageMap(path: string, paths: string[]) {
    if (!this._componentUsageMap[path])
      this._componentUsageMap[path] = new Set()

    paths.forEach((p) => {
      this._componentUsageMap[path].add(p)
    })
  }

  addComponents(paths: string | string[]) {
    debug.components('add', paths)

    const size = this._componentPaths.size
    toArray(paths).forEach(p => this._componentPaths.add(p))
    if (this._componentPaths.size !== size) {
      this.updateComponentNameMap()
      return true
    }
    return false
  }

  addCustomComponents(info: ComponentInfo) {
    if (info.name)
      this._componentCustomMap[info.name] = info
  }

  removeComponents(paths: string | string[]) {
    debug.components('remove', paths)

    const size = this._componentPaths.size
    toArray(paths).forEach(p => this._componentPaths.delete(p))
    if (this._componentPaths.size !== size) {
      this.updateComponentNameMap()
      return true
    }
    return false
  }

  onUpdate(path: string) {
    if (!this._server)
      return

    const payload: UpdatePayload = {
      type: 'update',
      updates: [],
    }
    const timestamp = +new Date()
    const name = pascalCase(getNameFromFilePath(path, this.options))

    Object.entries(this._componentUsageMap)
      .forEach(([key, values]) => {
        if (values.has(name)) {
          const r = `/${slash(relative(this.viteConfig.root, key))}`
          payload.updates.push({
            acceptedPath: r,
            path: r,
            timestamp,
            type: 'js-update',
          })
        }
      })

    if (payload.updates.length)
      this._server.ws.send(payload)

    this.generateDeclaration()
  }

  private updateComponentNameMap() {
    this._componentNameMap = {}

    Array
      .from(this._componentPaths)
      .forEach((path) => {
        const name = pascalCase(getNameFromFilePath(path, this.options))
        if (this._componentNameMap[name] && !this.options.allowOverrides) {
          // eslint-disable-next-line no-console
          console.warn(`[vite-plugin-components] component "${name}"(${path}) has naming conflicts with other components, ignored.`)
          return
        }

        this._componentNameMap[name] = {
          name,
          path: `/${this.relative(path)}`,
        }
      })
  }

  findComponent(name: string, excludePaths: string[] = [], rawName?: string): ComponentInfo | undefined {
    // resolve from fs
    let info = this._componentNameMap[name]
    if (info && !excludePaths.includes(info.path) && !excludePaths.includes(info.path.slice(1)))
      return info

    // custom resolvers
    for (const resolver of this.options.customComponentResolvers) {
      const result = resolver(name)
      if (result) {
        if (typeof result === 'string') {
          info = {
            name,
            path: result,
          }
          this.addCustomComponents(info)
          return info
        }
        else {
          info = {
            name,
            ...result,
          }
          this.addCustomComponents(info)
          return info
        }
      }
    }

    return undefined
  }

  findComponents(names: string[], excludePaths: string[] = []) {
    return names
      .map(name => this.findComponent(name, excludePaths))
      .filter(Boolean) as ComponentInfo[]
  }

  normalizePath(path: string) {
    // @ts-expect-error backward compatibility
    return resolveAlias(path, this.viteConfig?.resolve?.alias || this.viteConfig?.alias || [])
  }

  relative(path: string) {
    if (path.startsWith('/') && !path.startsWith(this.root))
      return slash(path.slice(1))
    return slash(relative(this.root, path))
  }

  _searched = false

  /**
   * This search for components in with the given options.
   * Will be called multiple times to ensure file loaded,
   * should normally run only once.
   *
   * @param ctx
   * @param force
   */
  searchGlob() {
    if (this._searched)
      return

    searchComponents(this)
    debug.search(this._componentNameMap)
    this._searched = true
  }

  generateDeclaration() {
    if (!this.options.globalComponentsDeclaration)
      return

    debug.decleration('generating')
    generateDeclaration(this, this.options.root, this.options.globalComponentsDeclaration)
  }

  get componentNameMap() {
    return this._componentNameMap
  }

  get componentCustomMap() {
    return this._componentCustomMap
  }
}
