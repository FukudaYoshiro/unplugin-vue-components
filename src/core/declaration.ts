import { resolve, dirname, relative } from 'path'
import { promises as fs, existsSync } from 'fs'
import { notNullish, slash } from '@antfu/utils'
import { Context } from './context'

export function parseDeclaration(code: string): Record<string, string> {
  return Object.fromEntries(Array.from(code.matchAll(/(?<!\/\/)\s+\s+['"]?(.+?)['"]?:\s(.+?)\n/g)).map(i => [i[1], i[2]]))
}

export async function generateDeclaration(ctx: Context, root: string, filepath: string) {
  const imports: Record<string, string> = Object.fromEntries(
    Object.values({
      ...ctx.componentNameMap,
      ...ctx.componentCustomMap,
    })
      .map(({ path, name, importName }) => {
        if (!name)
          return undefined

        const related = slash(path).startsWith('/')
          ? `./${relative(dirname(filepath), path)}`
          : path

        let entry = `typeof import('${slash(related)}')`
        if (importName)
          entry += `['${importName}']`
        else
          entry += '[\'default\']'
        return [name, entry]
      })
      .filter(notNullish),
  )

  if (!Object.keys(imports).length)
    return

  const originalImports = existsSync(filepath)
    ? parseDeclaration(await fs.readFile(filepath, 'utf-8'))
    : {}

  const lines = Object.entries({
    ...originalImports,
    ...imports,
  })
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(([name, v]) => {
      if (!/^\w+$/.test(name))
        name = `'${name}'`
      return `${name}: ${v}`
    })

  const code = `// generated by unplugin-vue-components
// We suggest you to commit this file into source control
// Read more: https://github.com/vuejs/vue-next/pull/3399

declare module 'vue' {
  export interface GlobalComponents {
    ${lines.join('\n    ')}
  }
}

export { }
`
  await fs.writeFile(filepath, code, 'utf-8')
}
