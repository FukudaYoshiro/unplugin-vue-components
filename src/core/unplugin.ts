import { createUnplugin } from 'unplugin'
import { Options, Transformer } from '../types'
import { Context } from './context'
import { parseId } from './utils'
import { Vue3Transformer } from './transforms/vue3'
import { Vue2Transformer } from './transforms/vue2'

export default createUnplugin<Options>((options = {}) => {
  const ctx: Context = new Context(options)
  let transformer: Transformer

  return {
    name: 'unplugin-vue-components',
    enforce: 'post',
    async transform(code, id) {
      const { path, query } = parseId(id)
      const result = await transformer(code, id, path, query)
      ctx.generateDeclaration()
      return result
    },

    vite: {
      configResolved(config) {
        ctx.setRoot(config.root)
        ctx.sourcemap = config.build.sourcemap

        if (config.plugins.find(i => i.name === 'vite-plugin-vue2'))
          options.transformer = options.transformer || 'vue2'

        transformer = ctx.options.transformer === 'vue2'
          ? Vue2Transformer(ctx)
          : Vue3Transformer(ctx)

        if (options.dts) {
          ctx.searchGlob()
          ctx.generateDeclaration()
        }
      },
      configureServer(server) {
        ctx.setViteServer(server)
      },
    },
  }
})
