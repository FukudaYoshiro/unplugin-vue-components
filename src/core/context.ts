import { relative } from 'path'
import Debug from 'debug'
import { UpdatePayload, ViteDevServer } from 'vite'
import { throttle, toArray, slash } from '@antfu/utils'
import { Options, ComponentInfo, ResolvedOptions } from '../types'
import { pascalCase, getNameFromFilePath, resolveAlias, matchGlobs } from './utils'
import { resolveOptions } from './options'
import { searchComponents } from './fs/glob'
import { generateDeclaration } from './declaration'

const debug = {
  components: Debug('unplugin-vue-components:context:components'),
  search: Debug('unplugin-vue-components:context:search'),
  hmr: Debug('unplugin-vue-components:context:hmr'),
  decleration: Debug('unplugin-vue-components:decleration'),
}

export class Context {
  options: ResolvedOptions

  private _componentPaths = new Set<string>()
  private _componentNameMap: Record<string, ComponentInfo> = {}
  private _componentUsageMap: Record<string, Set<string>> = {}
  private _componentCustomMap: Record<string, ComponentInfo> = {}
  private _server: ViteDevServer | undefined

  root = process.cwd()
  sourcemap: string | boolean = true
  alias: Record<string, string> = {}

  constructor(
    private rawOptions: Options,
  ) {
    this.options = resolveOptions(rawOptions, this.root)
    this.generateDeclaration = throttle(500, false, this.generateDeclaration.bind(this))
  }

  setRoot(root: string) {
    if (this.root === root)
      return
    this.root = root
    this.options = resolveOptions(this.rawOptions, this.root)
  }

  setViteServer(server: ViteDevServer) {
    this._server = server

    const { globs } = this.options

    server.watcher
      .on('unlink', (path) => {
        if (!matchGlobs(path, globs))
          return
        this.removeComponents(path)
        this.onUpdate(path)
      })
    server.watcher
      .on('add', (path) => {
        if (!matchGlobs(path, globs))
          return
        this.addComponents(path)
        this.onUpdate(path)
      })
  }

  /**
   * Record the usage of components
   * @param path
   * @param paths paths of used components
   */
  updateUsageMap(path: string, paths: string[]) {
    if (!this._componentUsageMap[path])
      this._componentUsageMap[path] = new Set()

    paths.forEach((p) => {
      this._componentUsageMap[path].add(p)
    })
  }

  addComponents(paths: string | string[]) {
    debug.components('add', paths)

    const size = this._componentPaths.size
    toArray(paths).forEach(p => this._componentPaths.add(p))
    if (this._componentPaths.size !== size) {
      this.updateComponentNameMap()
      return true
    }
    return false
  }

  addCustomComponents(info: ComponentInfo) {
    if (info.name)
      this._componentCustomMap[info.name] = info
  }

  removeComponents(paths: string | string[]) {
    debug.components('remove', paths)

    const size = this._componentPaths.size
    toArray(paths).forEach(p => this._componentPaths.delete(p))
    if (this._componentPaths.size !== size) {
      this.updateComponentNameMap()
      return true
    }
    return false
  }

  onUpdate(path: string) {
    if (!this._server)
      return

    const payload: UpdatePayload = {
      type: 'update',
      updates: [],
    }
    const timestamp = +new Date()
    const name = pascalCase(getNameFromFilePath(path, this.options))

    Object.entries(this._componentUsageMap)
      .forEach(([key, values]) => {
        if (values.has(name)) {
          const r = `/${slash(relative(this.root, key))}`
          payload.updates.push({
            acceptedPath: r,
            path: r,
            timestamp,
            type: 'js-update',
          })
        }
      })

    if (payload.updates.length)
      this._server.ws.send(payload)

    this.generateDeclaration()
  }

  private updateComponentNameMap() {
    this._componentNameMap = {}

    Array
      .from(this._componentPaths)
      .forEach((path) => {
        const name = pascalCase(getNameFromFilePath(path, this.options))
        if (this._componentNameMap[name] && !this.options.allowOverrides) {
          // eslint-disable-next-line no-console
          console.warn(`[unplugin-vue-components] component "${name}"(${path}) has naming conflicts with other components, ignored.`)
          return
        }

        this._componentNameMap[name] = {
          name,
          path: `/${this.relative(path)}`,
        }
      })
  }

  findComponent(name: string, excludePaths: string[] = [], rawName?: string): ComponentInfo | undefined {
    // resolve from fs
    let info = this._componentNameMap[name]
    if (info && !excludePaths.includes(info.path) && !excludePaths.includes(info.path.slice(1)))
      return info

    // custom resolvers
    for (const resolver of this.options.customComponentResolvers) {
      const result = resolver(name)
      if (result) {
        if (typeof result === 'string') {
          info = {
            name,
            path: result,
          }
          this.addCustomComponents(info)
          return info
        }
        else {
          info = {
            name,
            ...result,
          }
          this.addCustomComponents(info)
          return info
        }
      }
    }

    return undefined
  }

  findComponents(names: string[], excludePaths: string[] = []) {
    return names
      .map(name => this.findComponent(name, excludePaths))
      .filter(Boolean) as ComponentInfo[]
  }

  normalizePath(path: string) {
    // @ts-expect-error backward compatibility
    return resolveAlias(path, this.viteConfig?.resolve?.alias || this.viteConfig?.alias || [])
  }

  relative(path: string) {
    if (path.startsWith('/') && !path.startsWith(this.root))
      return slash(path.slice(1))
    return slash(relative(this.root, path))
  }

  _searched = false

  /**
   * This search for components in with the given options.
   * Will be called multiple times to ensure file loaded,
   * should normally run only once.
   *
   * @param ctx
   * @param force
   */
  searchGlob() {
    if (this._searched)
      return

    searchComponents(this)
    debug.search(this._componentNameMap)
    this._searched = true
  }

  generateDeclaration() {
    if (!this.options.dts)
      return

    debug.decleration('generating')
    generateDeclaration(this, this.options.root, this.options.dts)
  }

  get componentNameMap() {
    return this._componentNameMap
  }

  get componentCustomMap() {
    return this._componentCustomMap
  }
}
