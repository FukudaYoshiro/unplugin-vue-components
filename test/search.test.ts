import path from 'pathe'
import { describe, expect, it } from 'vitest'
import { Context } from '../src/core/context'

const root = path.resolve(__dirname, '../examples/vite-vue3')

function cleanup(data: any) {
  return Object.values(data).map((e: any) => {
    delete e.absolute
    e.from = path.relative(root, e.from).replace(/\\/g, '/')
    return e
  }).sort((a, b) => (a.as as string).localeCompare(b.as))
}

describe('search', () => {
  it('should work', async() => {
    const ctx = new Context({})
    ctx.setRoot(root)
    ctx.searchGlob()

    expect(cleanup(ctx.componentNameMap)).toMatchSnapshot()
  })

  it('should with namespace', async() => {
    const ctx = new Context({
      directoryAsNamespace: true,
      globalNamespaces: ['global'],
    })
    ctx.setRoot(root)
    ctx.searchGlob()

    expect(cleanup(ctx.componentNameMap)).toMatchSnapshot()
  })
})
