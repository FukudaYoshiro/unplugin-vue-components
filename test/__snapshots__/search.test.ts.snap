// Vitest Snapshot v1

exports[`search > should with namespace & collapse 1`] = `
[
  {
    "as": "Avatar",
    "from": "src/components/global/avatar.vue",
  },
  {
    "as": "Book",
    "from": "src/components/book/index.vue",
  },
  {
    "as": "CollapseFolderAndComponentPrefixes",
    "from": "src/components/collapse/collapseFolderAnd/CollapseFolderAndComponentPrefixes.vue",
  },
  {
    "as": "ComponentA",
    "from": "src/components/ComponentA.vue",
  },
  {
    "as": "ComponentAsync",
    "from": "src/components/ComponentAsync.vue",
  },
  {
    "as": "ComponentB",
    "from": "src/components/ComponentB.vue",
  },
  {
    "as": "ComponentC",
    "from": "src/components/component-c.vue",
  },
  {
    "as": "ComponentD",
    "from": "src/components/ComponentD.vue",
  },
  {
    "as": "Recursive",
    "from": "src/components/Recursive.vue",
  },
  {
    "as": "UiButton",
    "from": "src/components/ui/button.vue",
  },
  {
    "as": "UiNestedCheckbox",
    "from": "src/components/ui/nested/checkbox.vue",
  },
]
`;

exports[`search > should with namespace 1`] = `
[
  {
    "as": "Avatar",
    "from": "src/components/global/avatar.vue",
  },
  {
    "as": "Book",
    "from": "src/components/book/index.vue",
  },
  {
    "as": "CollapseCollapseFolderAndCollapseFolderAndComponentPrefixes",
    "from": "src/components/collapse/collapseFolderAnd/CollapseFolderAndComponentPrefixes.vue",
  },
  {
    "as": "ComponentA",
    "from": "src/components/ComponentA.vue",
  },
  {
    "as": "ComponentAsync",
    "from": "src/components/ComponentAsync.vue",
  },
  {
    "as": "ComponentB",
    "from": "src/components/ComponentB.vue",
  },
  {
    "as": "ComponentC",
    "from": "src/components/component-c.vue",
  },
  {
    "as": "ComponentD",
    "from": "src/components/ComponentD.vue",
  },
  {
    "as": "Recursive",
    "from": "src/components/Recursive.vue",
  },
  {
    "as": "UiButton",
    "from": "src/components/ui/button.vue",
  },
  {
    "as": "UiNestedCheckbox",
    "from": "src/components/ui/nested/checkbox.vue",
  },
]
`;

exports[`search > should work 1`] = `
[
  {
    "as": "Avatar",
    "from": "src/components/global/avatar.vue",
  },
  {
    "as": "Book",
    "from": "src/components/book/index.vue",
  },
  {
    "as": "Button",
    "from": "src/components/ui/button.vue",
  },
  {
    "as": "Checkbox",
    "from": "src/components/ui/nested/checkbox.vue",
  },
  {
    "as": "CollapseFolderAndComponentPrefixes",
    "from": "src/components/collapse/collapseFolderAnd/CollapseFolderAndComponentPrefixes.vue",
  },
  {
    "as": "ComponentA",
    "from": "src/components/ComponentA.vue",
  },
  {
    "as": "ComponentAsync",
    "from": "src/components/ComponentAsync.vue",
  },
  {
    "as": "ComponentB",
    "from": "src/components/ComponentB.vue",
  },
  {
    "as": "ComponentC",
    "from": "src/components/component-c.vue",
  },
  {
    "as": "ComponentD",
    "from": "src/components/ComponentD.vue",
  },
  {
    "as": "Recursive",
    "from": "src/components/Recursive.vue",
  },
]
`;
